src/Dapr.Messaging/PublishSubscribe/DaprPublishSubscribeGrpcClient.cs

```diff
@@ -0,0 +1,94 @@
+﻿// ------------------------------------------------------------------------
+// Copyright 2024 The Dapr Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ------------------------------------------------------------------------
+
+using P = Dapr.Client.Autogen.Grpc.v1.Dapr;
+
+namespace Dapr.Messaging.PublishSubscribe;
+
+/// <summary>
+/// A client for interacting with the Dapr endpoints.
+/// </summary>
+internal sealed class DaprPublishSubscribeGrpcClient : DaprPublishSubscribeClient
+{
+    /// <summary>
+    /// The various receiver clients created for each combination of Dapr pubsub component and topic name.
+    /// </summary>
+    private readonly Dictionary<(string, string), PublishSubscribeReceiver> clients = new();
+    /// <summary>
+    /// The Dapr client.
+    /// </summary>
+    private readonly P.DaprClient daprClient;
+
+    // property exposed for testing purposes
+    internal P.DaprClient Client => daprClient;
+
+    /// <summary>
+    /// Creates a new instance of a <see cref="DaprPublishSubscribeGrpcClient"/>
+    /// </summary>
+    public DaprPublishSubscribeGrpcClient(P.DaprClient client)
+    {
+        daprClient = client;
+    }
+
+    /// <summary>
+    /// Dynamically subscribes to a Publish/Subscribe component and topic.
+    /// </summary>
+    /// <param name="pubsubName">The name of the Publish/Subscribe component.</param>
+    /// <param name="topicName">The name of the topic to subscribe to.</param>
+    /// <param name="options">Configuration options.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>An <see cref="IAsyncEnumerable{TopicMessage}"/> containing the various messages returned by the subscription.</returns>
+    public override IAsyncEnumerable<TopicMessage> SubscribeAsync(string pubsubName, string topicName, DaprSubscriptionOptions options,
+        CancellationToken cancellationToken)
+    {
+        var receiver = new PublishSubscribeReceiver(pubsubName, topicName, options, daprClient);
```

Comment:
Do we need to guard against concurrent/duplicate subscriptions to the same pubsub/topics?

Reply:
Good call - I think so. If this later supports some sort of partitioning functionality to allow multiple subscribers to access the same resource, we can revisit it, but I've added a guard against this so existing receivers will be re-used.

Reply:
The approach I've taken here is simple - allow only one subscriber per topic name and pubsub component name and throw if a duplicate subscription is made. 

We can revisit this to allow multiple subscriptions per combination down the line, but the biggest difficulty in enabling that is lack of consumer group/partition support on the Dapr runtime side. Put simply, having more than one consumer of the same topic will open the door to consumers potentially getting out of sync on deliveries and potentially returning different acknowledgements for the same message identifiers. I propose we save that problem for another day and opt for the easier approach for now. 

Should developers want to support multiple subscribers, they need only persist the messages to an internal queue/channel from the delegate and pull from it how they want (as it wouldn't need acknowledgement support at that downstream level) to easily work around this SDK restriction.

Reply:
Revisited this later on - no such restriction any longer.